### PR TITLE
RavenDB-6440 - Implement attachments replication - WIP

### PR DESCRIPTION
**Pull request by fitzchak**

### CODE REVIEW

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 1]
```diff
@@ -290,7 +290,12 @@ private void HandleReceivedIndexOrTransformerBatch(TransactionOperationContext c
         private void HandleReceivedDocumentBatch(DocumentsOperationContext documentsContext, BlittableJsonReaderObject message, long lastDocumentEtag)
         {
             int itemCount;
-            if (!message.TryGet(nameof(ReplicationMessageHeader.ItemCount), out itemCount))
+            if (!message.TryGet(nameof(ReplicationMessageHeader.ItemsCount), out itemCount))
+                throw new InvalidDataException(
+                    "Expected the 'ItemCount' field, but had no numeric field of this value, this is likely a bug");
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:295]: Need to fix the name here.

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 2]
```diff
@@ -290,7 +290,12 @@ private void HandleReceivedIndexOrTransformerBatch(TransactionOperationContext c
...
+
+            int attachmentStreamCount;
+            if (!message.TryGet(nameof(ReplicationMessageHeader.AttachmentStreamsCount), out attachmentStreamCount))
                 throw new InvalidDataException(
                     "Expected the 'ItemCount' field, but had no numeric field of this value, this is likely a bug");
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:300]: Invalid message here.

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 3]
```diff
@@ -794,6 +802,15 @@ public struct ReplicationDocumentsPositions
             public DocumentFlags Flags;
         }
 
+        public struct ReplicationAttachment
+        {
+            public short TransactionMarker;
+            public byte[] LoweredKey;
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:808]: What are we using managed arrays here?

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 4]
```diff
@@ -810,17 +827,54 @@ public struct ReplicationIndexOrTransformerPositions
...
+                var type = *(ReplicationBatchItem.ReplicationItemType*)ReadExactly(sizeof(byte));
+                if (type == ReplicationBatchItem.ReplicationItemType.Attachment)
+                {
+                    var attachment = new ReplicationAttachment
+                    {
+                        TransactionMarker = *(short*)ReadExactly(sizeof(short)),
+                    };
+
+                    var loweredKeySize = *(int*)ReadExactly(sizeof(int));
+                    attachment.LoweredKey = new byte[loweredKeySize];
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:842]: Need validation for size that is too large

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 5]
```diff
@@ -1046,19 +1099,25 @@ public override void Execute(DocumentsOperationContext context)
                         document?.Dispose();
                     }
                 }
+
+                foreach (var attachment in _incoming._replicatedAttachments)
+                {
+                    if (_incoming._log.IsInfoEnabled)
+                        _incoming._log.Info($"Got incoming attachment, doing PUT on attachment = {attachment.Name}, with key = {Encoding.UTF8.GetString(attachment.LoweredKey)}");
+                    database.DocumentsStorage.AttachmentsStorage.PutFromReplication(context, ref attachment);
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1107]: Why are we using a ref here?

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 1 above)
Comment (fitzchak) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:295]: Done

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 2 above)
Comment (fitzchak) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:300]: Fixed

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 4 above)
Comment (fitzchak) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:842]: We have it here:
https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs#L409

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 3 above)
Comment (fitzchak) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:808]: This should be changed.

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 5 above)
Comment (fitzchak) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1107]: This should have been changed. This method is WIP anyway.

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:842]: This is on the send, this is on the read

**Pull request merged**